File: May-LeetCoding-Challenge/15-Maximum-Sum-Circular-Subarray/Maximum-Sum-Circular-Subarray.cpp
class Solution {
public:
    int maxSubarraySumCircular(const vector<int>& A) {
        int max_sum = A[0], min_sum = A[0];
        int sum1 = 0, sum2 = 0, tot = 0;

        for (auto &x: A) {
            sum1 += x; 
            sum2 += x;
            tot += x;
            if (sum1 > max_sum) max_sum = sum1;
            if (sum2 < min_sum) min_sum = sum2;
            if (sum1 < 0) sum1 = 0;
            if (sum2 > 0) sum2 = 0;
        }

        return (tot == min_sum)?max_sum:max(max_sum, tot - min_sum);
    }
};

File: May-LeetCoding-Challenge/15-Maximum-Sum-Circular-Subarray/Maximum-Sum-Circular-Subarray.kt
class MaximumSumCircularSubarrayKotlin918 {
    // Kadane’s Algorithm
    /*
    if maxSum array in [0 until A.size] -> maxSum
    if maxSum array in [..A.size -- 0 ..] -> sumOfArray - minSum
     */
    fun maxSubarraySumCircular(A: IntArray): Int {
        if (A.isEmpty()) {
            return 0
        }
        var sum = 0
        var maxSum = Int.MIN_VALUE
        var minSum = Int.MAX_VALUE
        var currentMax = 0
        var currentMin = 0
        A.forEach {
            currentMax = maxOf(it, currentMax + it)
            maxSum = maxOf(maxSum, currentMax)
            currentMin = minOf(it, currentMin + it)
            minSum = minOf(minSum, currentMin)
            sum += it
        }
        // if all element < 0, return maxSum
        return when {
            maxSum < 0 -> maxSum
            else -> maxOf(maxSum, sum - minSum)
        }
    }
}

fun main() {
    val solution = MaximumSumCircularSubarrayKotlin918()
    println(solution.maxSubarraySumCircular(intArrayOf(1, -2, 3, -2))) // 3
    println(solution.maxSubarraySumCircular(intArrayOf(5, -3, 5))) // 10
    println(solution.maxSubarraySumCircular(intArrayOf(3, -1, 2, -1))) // 4
    println(solution.maxSubarraySumCircular(intArrayOf(3, -2, 2, -3))) // 3
    println(solution.maxSubarraySumCircular(intArrayOf(-3, -2, -5, -8))) // -2
}